runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -235,32 +261,60 @@ public int appendObjectState(Stream<Map.Entry> entryStream) {
         ImmutableMap<CheckpointEntry.CheckpointDictKey, String> mdkv =
                 ImmutableMap.copyOf(this.mdkv);
 
-        final Iterable<List<Map.Entry>> partitions =
-                Iterables.partition(entryStream::iterator, batchSize);
-        int entryCount = 0;
-
-        for (List<Map.Entry> partition : partitions) {
-            MultiSMREntry smrEntries = new MultiSMREntry();
-            for (Map.Entry entry : partition) {
-                smrEntries.addTo(new SMREntry("put",
-                        new Object[]{keyMutator.apply(entry.getKey()),
-                                valueMutator.apply(entry.getValue())},
-                        serializer));
-                entryCount++;
+        int totalEntryCount = 0;
+        int currentBatchSize = 0;
+        int numBytesPerCheckpointEntry = 0;
+        int maxWriteSizeLimit = (int) (BATCH_THRESHOLD_PERCENTAGE * rt.getParameters().getMaxWriteSize());
+
+        ByteBuf inputBuffer = Unpooled.buffer();
+        MultiSMREntry smrEntries = new MultiSMREntry();
+
+        for (Map.Entry entry : (Iterable<Map.Entry>) entryStream::iterator) {
+            SMREntry smrPutEntry = new SMREntry("put",
+                    new Object[]{keyMutator.apply(entry.getKey()),
+                            valueMutator.apply(entry.getValue())},
+                    serializer);
+
+            /* Need to check the size of the compressed buffer. inputByteBuffer and
+               inputBuffer are the same buffer.
+             */
+            smrPutEntry.serialize(inputBuffer);
+            ByteBuffer compressedBuffer =
+                    rt.getParameters()
+                            .getCodecType()
+                            .getInstance().compress(ByteBuffer.wrap(inputBuffer.array(),
+                            inputBuffer.readerIndex(), inputBuffer.writerIndex()));
+
+            numBytesPerCheckpointEntry += compressedBuffer.limit();
+
+            /* this is a conservative approach, size limit is calculated on uncompressed data
```

Comment:
this comment is outdated, correct? Now it seems the calculation is made on compressed data...

Reply:
yes, the calculation is made on compressed data